jupyterhub_moss/form/option_form.js

```diff
@@ -68,6 +77,83 @@ function setSimplePartition(name) {
   };
 }
 
+function isCustomEnvironment() {
+  const environmentElem = document.getElementById('environment');
+  const index = environmentElem.selectedIndex;
+
+  return index !== -1 && environmentElem.options[index].id === 'environment_custom';
+}
+
+function getEnvironmentPath() {
+  const environmentElem = document.getElementById('environment');
+  const environmentCustomOptionElem = document.getElementById('environment_custom');
+
+  const partition = document.getElementById('partition').value;
+  const info = window.SLURM_DATA.partitions[partition];
+
+  return isCustomEnvironment() ? environmentCustomOptionElem.value : info.jupyter_environments[environmentElem.value]['path'];
+}
+
+function updateEnvironmentPath() {
+  const environmentPathElem = document.getElementById('environment_path');
+  const environmentPathNoteElem = document.getElementById('environment_path_note');
+
+  setReadOnly(environmentPathElem, !isCustomEnvironment());
+  environmentPathElem.style.borderColor = isCustomEnvironment() ? null : 'transparent';
```

Comment:
Use the text input for both pre-defined env. path and custom one by toggling read-only and border.